### PR TITLE
[BUGFIX DEPRECATE] fixes and deprecates when a user re-exports a model multiple places for use

### PR DESCRIPTION
**Pull request by runspired**
@bmac If good this should be backported to release and beta.

This twiddle demonstrates the issue: https://ember-twiddle.com/035a072e1818f48c09ac9a38f13b1773?openFiles=twiddle.json%2C

All Ember versions prior to 2.12 work with all ED versions. Beginning in 2.12 (I think we fussed with double extend) this begins breaking in ED.

We can temporarily fix this with a deprecation by extending the class ourselves.

### CODE REVIEW

**Comment by runspired** [comment]
I like the suggestion from @hjdivad. Since @stefanpenner is out, perhaps @krisselden or @rwjblue has thoughts on the best approach here?

**Comment by runspired** [comment]
@hjdivad updated

File: addon/-private/system/store.js [hunk 1]
```diff
@@ -2107,13 +2107,40 @@ Store = Service.extend({
...
+      if (klass.modelName && klass.modelName !== modelName) {
+        const Message =
+          `Ember Data found a Model for ${modelName} but the class ` +
+          `already had the modelName ${klass.modelName}.\n\n` +
+          'This likely means that you are re-exporting a Model.\n\n' +
+          `Replace:\n\texport { default } from './${klass.modelName}';\nwith:\n\t` +
+          `import Model from './${klass.modelName}';\n\n\texport default Model.extend();`;
+
+        if (hasFactoryFor) {
+          klass = factory.class = klass.extend();
```
Comment (rwjblue) [inline addon/-private/system/store.js]: It seems like a bug to me that ember allows setting factory.class like this...

Comment (runspired) [inline addon/-private/system/store.js]: if it is, then I don't think we have a path forward.

Comment (krisselden) [inline addon/-private/system/store.js]: factoryFor isn’t supposed to allow that, I thought we were using proxies to error in debug for that.

Comment (rwjblue) [inline addon/-private/system/store.js]: @krisselden - We do exactly that, for all other properties other than `create` and `class`. We should fix this (e.g. throw if class or create are attempted to be set).

Comment (rwjblue) [inline addon/-private/system/store.js]: @runspired - We could still do what you want I think. Either:

* do the double extend when we detect modelName is not a match, register, then look up again for them...
* create an cache our own "fake factory" here...

Comment (runspired) [inline addon/-private/system/store.js]: I think I prefer the register approach given it is "more correct" in terms of intent. Would be weird if ED and the registry were out of sync.

Comment (runspired) [inline addon/-private/system/store.js]: cc @rwjblue

Comment (stefanpenner) [inline addon/-private/system/store.js]: @rwjblue / @krisselden i totally agree with you guys, the issue is ember-data's reliance on `constructor.modelName`. I have removed as much of this dependence as can (I believe) without public API change...

File: addon/-private/system/store.js [hunk 2]
```diff
@@ -2107,13 +2107,40 @@ Store = Service.extend({
...
-      // TODO: deprecate this
-      klass.modelName = klass.modelName || modelName;
+      // TODO: deprecate klass.modelName
+      /*
+        In test environments it is common to re-use the same base model in multiple
+        tests and register it each time. Since that seems convenient and good
+        to continue allowing this ignores the case where modelName is set but is
+         the same.
+       */
+      if (klass.modelName && klass.modelName !== modelName) {
```
Comment (stefanpenner) [inline addon/-private/system/store.js]: In theory (for the test case), we could detect if the original container has been destroyed or changed and if so, just seamlessly transition without a warning?

File: addon/-private/system/store.js (hunk 1 above)
Comment (stefanpenner) [inline addon/-private/system/store.js]: I believe if users have tests that pollute the constructor, they should subclass in the own tests. I don't believe we should subclass for them.

File: addon/-private/system/store.js (hunk 2 above)
Comment (runspired) [inline addon/-private/system/store.js]: ugh, I hadn't even considered tests mutating this yet. :'( 

Thoughts on how would we differentiate between "modelName set by previous container" and "modelName set by previous location's lookup"?

Comment (stefanpenner) [inline addon/-private/system/store.js]: if the `owner` is different? Maybe we store an owner version.

---

* I think tests should just double extend themselves (this is aligned with how the language works)
* we should just explicitly not support double re-exports of models until `modelName` goes away entirely.

**Comment by runspired** [comment]
I believe this is what folks are hitting in #6143

**Pull request closed without merging**